### PR TITLE
Fix issue with top bar menu having unique ARIA roles

### LINKED ISSUE #12505: The top bar More Menu should be a unique ARIA menu
The More Menu in the top toolbar is made of a collection of ARIA menus, each one with its own menu items. Plugins can add their own menu items in the "Plugins" sub-menu.

However, this menu (screenshot on the left) should work in the same way as the Block "More options" menu (screenshot on the right):

screenshot 2018-12-01 at 15 51 38

I'm not referring to the visuals: it's about semantics, accessibility, and keyboard interaction:
- the whole menu should be a single ARIA menu 
- groups can be made using an ARIA `role="group"`, see the W3C example from the ARIA authoring practices: https://www.w3.org/TR/wai-aria-practices-1.1/examples/menubar/menubar-2/menubar-2.html where "Style/Color" and "Size" have groups
- each group can use a separator and should be labelled with `aria-labelledby` referencing the visible title
- most importantly: this way, arrows navigation will work across the whole menu, while currently it only works within each single group

Basically, the `MenuGroup` component should be just a `group` of menu items and not render a `NavigableMenu`. Then, the groups should be wrapped by an ARIA menu, i.e. a single `NavigableMenu`.

Some care should be used for the visible "titles", as the items container in an ARIA menu can only be menuitems or separators child elements of the containing menu so the title should be removed from the accessibility tree.

This will probably require some small refactoring and evaluating how the changes impact the current usage of these components.

Reference:
https://www.w3.org/TR/wai-aria-practices-1.1/#menu

### PR DESCRIPTION
**Pull request by Firestorm980**
## Description
Closes https://github.com/WordPress/gutenberg/issues/12505. Adds role options to menus so that the items within the main more menu can now be groups within the menu. Should fix issues with keyboard navigation as well as roles. The menu options/props default to what they used to be, so if it isn't specified and these menus are used elsewhere, it should fallback to the old functionality.

## How has this been tested?

- Manual keyboard navigation testing in the admin with VoiceOver.
- Ran `test-e2e` within Gutenberg docker.

## Types of changes

- Updates menu group component to have a default role, label, and keyboard event control
- Changes menus used in the top nav to use `group` instead of `menu` and wraps within a parent `menu`
- Changes menus used in the top nav to let the parent menu handle overall keyboard event navigation

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/components/src/menu-group/index.js [hunk 1]
```diff
@@ -19,6 +19,10 @@ export function MenuGroup( {
 	className = '',
 	instanceId,
 	label,
+	role = 'menu',
+	labelRole = ( role === 'group' ) ? { role: 'presentation' } : {},
+	useEventToOffset = true,
```
Comment (jorgefilipecosta) [inline packages/components/src/menu-group/index.js]: We already have an eventToOffset to offset property, any reason for using a different name here  (useEventToOffset)?

File: packages/components/src/menu-group/index.js [hunk 2]
```diff
@@ -33,9 +37,14 @@ export function MenuGroup( {
 	return (
 		<div className={ classNames }>
 			{ label &&
-				<div className="components-menu-group__label" id={ labelId }>{ label }</div>
+				<div
+					className={ `components-menu-group__label ${ ( isScreenReaderLabel ) ? 'screen-reader-text' : '' }` }
```
Comment (jorgefilipecosta) [inline packages/components/src/menu-group/index.js]: Here if isScreenReaderLabel is false we add an unnecessary space. What if we use the classnames util e.g:
```
className={ classnames( 'components-menu-group__label', {
['screen-reader-text']: isScreenReaderLabel
} }
```

**Review by jorgefilipecosta — COMMENTED** [review]
Hi @Firestorm980 thank you for your contribution. I don't notice any visual changes and it looks like things are working correctly.
Previously on the ellipsis menu, it was not possible to use the arrow down to navigate between all the menu items and now it is (which is an improvament).
cc: @afercia, @tofumatt would it be possible to have a look at this PR regarding the a11y point of view?

File: packages/components/src/menu-group/index.js (hunk 1 above)
Comment (Firestorm980) [reply packages/components/src/menu-group/index.js]: @jorgefilipecosta I felt it was clearer in this context, but if we prefer consistency, I can change that.

**Comment by gziolo** [comment]
@Firestorm980 - do you plan to address feedback left? It looks like it is very close to be ready to ship. We would like to include it in the plugin's release planned for the middle of February.

**Comment by Firestorm980** [comment]
@gziolo I should be able to get to this one in the next couple days.

**Comment by Firestorm980** [comment]
@gziolo @afercia Fixed should be included now.

File: packages/components/src/menu-group/index.js [hunk 3]
```diff
@@ -19,6 +19,10 @@ export function MenuGroup( {
 	className = '',
 	instanceId,
 	label,
+	role = 'menu',
+	labelRole = ( role === 'group' ) ? { role: 'presentation' } : {},
```
Comment (afercia) [inline packages/components/src/menu-group/index.js]: I'd remove this, as role=presentation is useful to "flatten" the accessibility tree but there's no need for it, as the `div` it applies to is already ignored.

File: packages/components/src/menu-group/index.js [hunk 4]
```diff
@@ -19,6 +19,10 @@ export function MenuGroup( {
...
+	useEventToOffset = true,
+	isScreenReaderLabel = false,
```
Comment (afercia) [inline packages/components/src/menu-group/index.js]: not sure we need to visually hide the group "label", as groups with only one item shouldn't be groups (see below)

File: packages/components/src/menu-group/index.js [hunk 5]
```diff
@@ -33,9 +37,16 @@ export function MenuGroup( {
 	return (
 		<div className={ classNames }>
 			{ label &&
-				<div className="components-menu-group__label" id={ labelId }>{ label }</div>
+				<div
+					className={ classnames( 'components-menu-group__label', {
+						'screen-reader-text': isScreenReaderLabel,
+					} ) }
+					{ ...labelRole }
```
Comment (afercia) [inline packages/components/src/menu-group/index.js]: See above.

File: packages/components/src/menu-group/index.js [hunk 6]
```diff
@@ -33,9 +37,16 @@ export function MenuGroup( {
 	return (
 		<div className={ classNames }>
 			{ label &&
-				<div className="components-menu-group__label" id={ labelId }>{ label }</div>
+				<div
+					className={ classnames( 'components-menu-group__label', {
+						'screen-reader-text': isScreenReaderLabel,
+					} ) }
```
Comment (afercia) [inline packages/components/src/menu-group/index.js]: see above

File: packages/edit-post/src/components/header/more-menu/index.js [hunk 7]
```diff
@@ -33,11 +34,12 @@ const MoreMenu = () => (
 		) }
 		renderContent={ ( { onClose } ) => (
 			<Fragment>
-				<WritingMenu onClose={ onClose } />
-				<ModeSwitcher onSelect={ onClose } />
-				<PluginMoreMenuGroup.Slot fillProps={ { onClose } } />
-				<ToolsMoreMenuGroup.Slot fillProps={ { onClose } } />
-				<MenuGroup>
+				<MenuGroup label={ menuLabel } isScreenReaderLabel={ true }>
```
Comment (afercia) [inline packages/edit-post/src/components/header/more-menu/index.js]: this parent group should hide its label though

File: packages/edit-post/src/components/header/mode-switcher/index.js [hunk 8]
```diff
@@ -36,6 +36,8 @@ function ModeSwitcher( { onSwitch, mode } ) {
 
 	return (
 		<MenuGroup
+			role={ 'group' }
```
Comment (afercia) [inline packages/edit-post/src/components/header/mode-switcher/index.js]: I guess this could be just role="group"

File: packages/edit-post/src/components/header/options-menu-item/index.js [hunk 9]
```diff
@@ -3,18 +3,24 @@
...
+			isScreenReaderLabel={ true }>
+			<MenuItem
+				onClick={ () => {
+					onSelect();
+					openModal( 'edit-post/options' );
+				} }
+			>
+				{ __( 'Options' ) }
+			</MenuItem>
+		</MenuGroup>
```
Comment (afercia) [inline packages/edit-post/src/components/header/options-menu-item/index.js]: I wouldn't change this in a group: it's a single menu item within the main parent group. Also, it doesn't need a label that just repeats the name of the item. I'd just revert all the changes to this file.

File: packages/edit-post/src/components/header/plugins-more-menu-group/index.js [hunk 10]
```diff
@@ -14,7 +14,10 @@ const { Fill: PluginsMoreMenuGroup, Slot } = createSlotFill( 'PluginsMoreMenuGro
 PluginsMoreMenuGroup.Slot = ( { fillProps } ) => (
 	<Slot fillProps={ fillProps }>
 		{ ( fills ) => ! isEmpty( fills ) && (
-			<MenuGroup label={ __( 'Plugins' ) }>
+			<MenuGroup
+				role={ 'group' }
```
Comment (afercia) [inline packages/edit-post/src/components/header/plugins-more-menu-group/index.js]: I guess this could be just role="group"

File: packages/edit-post/src/components/header/tools-more-menu-group/index.js [hunk 11]
```diff
@@ -14,7 +14,10 @@ const { Fill: ToolsMoreMenuGroup, Slot } = createSlotFill( 'ToolsMoreMenuGroup'
 ToolsMoreMenuGroup.Slot = ( { fillProps } ) => (
 	<Slot fillProps={ fillProps }>
 		{ ( fills ) => ! isEmpty( fills ) && (
-			<MenuGroup label={ __( 'Tools' ) }>
+			<MenuGroup
+				role={ 'group' }
```
Comment (afercia) [inline packages/edit-post/src/components/header/tools-more-menu-group/index.js]: I guess this could be just role="group"

File: packages/edit-post/src/components/header/writing-menu/index.js [hunk 12]
```diff
@@ -13,6 +13,8 @@ import FeatureToggle from '../feature-toggle';
 function WritingMenu( { onClose } ) {
 	return (
 		<MenuGroup
+			role={ 'group' }
```
Comment (afercia) [inline packages/edit-post/src/components/header/writing-menu/index.js]: I guess this could be just role="group"

**Comment by afercia** [comment]
Left a few comments. Basically there's the need to get a structure of the accessibility tree where the main menu contains only groups and single items. The tree can be "flattened" by using role=presentation or even aria-hidden=true where necessary. This kind of information can be inspected in the Chrome dev tools > Elements > Accessibility tab, where each element can be checked and it's also possible to check which elements are "ignored".

Current tree:

screenshot 2019-02-16 at 19 48 54

Desired tree:

screenshot 2019-02-16 at 20 14 23

**Comment by afercia** [comment]
@Firestorm980 hello. Do you have time to address the pending issues? If not, please let us know. These changes would greatly improve keyboard interaction in the more menu. If you don't have time, we can always copy your branch to the main repo preserving commits and attributions, and proceed from there. Thanks!

**Comment by gziolo** [comment]
@afercia in case we don't hear back from @Firestorm980 soon, I opened #14525 whit all commits transferred. In addition, I rebased new PR with all the changes applied in `master` branch. Feel free to work on new PR.

File: packages/edit-post/src/components/header/mode-switcher/index.js (hunk 8 above)
Comment (gziolo) [reply packages/edit-post/src/components/header/mode-switcher/index.js]: Yes, correct.

**Comment by chrisvanpatten** [comment]
Removed the "stale" label as it seems it no longer is :)

**Comment by Firestorm980** [comment]
Sorry about the delay in communication on this one. Hopefully will have another commit up soon with fixes.

File: packages/components/src/navigable-container/container.js [hunk 13]
```diff
@@ -68,6 +68,11 @@ class NavigableContainer extends Component {
 		const { getFocusableContext } = this;
 		const { cycle = true, eventToOffset, onNavigate = noop, stopNavigationEvents } = this.props;
```
Comment (gziolo) [inline packages/components/src/navigable-container/container.js]:
```suggestion
		const { cycle = true, eventToOffset = noop, onNavigate = noop, stopNavigationEvents } = this.props;
```

This should be enough to make this work instead of the check:
> // This avoids a console error when eventToOffset is null

File: packages/edit-post/src/components/header/more-menu/index.js [hunk 14]
```diff
@@ -33,11 +34,12 @@ const MoreMenu = () => (
 		) }
 		renderContent={ ( { onClose } ) => (
 			<Fragment>
-				<WritingMenu onClose={ onClose } />
-				<ModeSwitcher onSelect={ onClose } />
-				<PluginMoreMenuGroup.Slot fillProps={ { onClose } } />
-				<ToolsMoreMenuGroup.Slot fillProps={ { onClose } } />
-				<MenuGroup>
+				<MenuGroup label={ menuLabel } isLabelHidden={ true }>
```
Comment (gziolo) [inline packages/edit-post/src/components/header/more-menu/index.js]: `Fragment` is no longer necessary to wrap components.

File: packages/components/src/menu-group/index.js [hunk 15]
```diff
@@ -19,6 +19,9 @@ export function MenuGroup( {
 	className = '',
 	instanceId,
 	label,
+	role = 'menu',
```
Comment (gziolo) [inline packages/components/src/menu-group/index.js]: It feels wrong to see the default role set `menu` for the component which is called `MenuGroup` and then see it overridden back to `group` in other places. It seems like we need to introduce another component called `Menu` which will have `role` set to `menu`, its class name which will hide the label for the menu, and will always have `useEventToOffset` set to `true`. This will make it possible to leave `MenuGroup` usage unchanged.

**Review by gziolo — CHANGES_REQUESTED** [review]
See my comments. In general, this PR goes in the right direction but it would be nice to explore whether we should introduce new common component `Menu` which represents the modified version of `MenuGroup`. While those two are similar, they are conceptually different.

File: packages/components/src/menu-group/index.js (hunk 15 above)
Comment (timwright12) [reply packages/components/src/menu-group/index.js]: @gziolo should we spin up a new issue for that?

Comment (gziolo) [reply packages/components/src/menu-group/index.js]: I would feel comfortable doing it as part of this PR because in the current form you might see unexpected `role="menu"` in all plugins using `MenuGroup`.

**Comment by aduth** [comment]
> In general, this PR goes in the right direction but it would be nice to explore whether we should introduce new common component `Menu` which represents the modified version of `MenuGroup`. While those two are similar, they are conceptually different.

>This will make it possible to leave `MenuGroup` usage unchanged.

Arriving here via #14843 and considering the recommendations from #12505, I wonder what it is that `MenuGroup` is supposed to represent that we're trying to leave untouched. From [its current documentation](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/menu-group), it's not clear to me why it implements behaviors (navigating by arrow keys) and not instead rely on that to be handled by the context in which it's rendered. Indeed, this seems to be the recommendation in #12505:

>Basically, the `MenuGroup` component should be just a `group` of menu items and not render a `NavigableMenu`. Then, the groups should be wrapped by an ARIA menu, i.e. a single `NavigableMenu`.

The challenge is in how it's used today, the "More Options" menu doesn't implement any of these behaviors on the wrapping dropdown. It seems like something we'd expect out of the [`DropdownMenu` component](https://github.com/WordPress/gutenberg/tree/master/packages/components/src/dropdown-menu), but as it exists today, it's not possible to actually use groupings there. #14843 is a step in the direction of making this possible, though.

I would expect:

- The "More Options" menu implements using `NavigableMenu`, probably where today [it renders just a `Fragment`](https://github.com/WordPress/gutenberg/blob/1df302121f0ab7df8611d01e28ad41db25f40d52/packages/edit-post/src/components/header/more-menu/index.js#L35)
- Each of the sections ("Writing", etc) render as their own `MenuGroup`.
- A `MenuGroup` is just a `div` with grouping attributes as prescribed in #12505 (i.e. no `NavigableMenu`).
- After #14843, "More Options" component could be refactored to use `DropdownMenu` instead of `Dropdown` so that it doesn't need to handle `NavigableMenu` behaviors. 

tl;dr We seem to be trying to let `MenuGroup` to have some behaviors for a standalone context, and I'm not sure why?

**Comment by Firestorm980** [comment]
@gziolo @aduth just a heads up that my schedule is getting pretty full so I don't have a lot of spare time to tackle this one, especially now that the scope of the ticket appears to be changing somewhat.

**Comment by gziolo** [comment]
> @gziolo @aduth just a heads up that my schedule is getting pretty full so I don't have a lot of spare time to tackle this one, especially now that the scope of the ticket appears to be changing somewhat.

Thank you for your contribution so far, it helped us better understand what would be the best approach here. I'm personally leaning towards including the changes proposed here in #14843. It means there is going to be more aggressive refactoring required to make it work as @afercia described in #12505. However, it seems like we would be able to address all concerns shared in the original issue. In addition, this will give us more control over `DropdownMenu` and will improve the default usage of this component in the context of its accessibility.
